ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+      srcIndex.hardDeleter.resume();
+    }
+    isActive = true;
+  }
+
+  /**
+   * Closes the compactor and waits for {@code waitTimeSecs} for the close to complete.
+   * @param waitTimeSecs the number of seconds to wait for close to complete.
+   * @throws InterruptedException if the wait for close was interrupted.
+   */
+  void close(long waitTimeSecs) throws InterruptedException {
+    isActive = false;
+    if (waitTimeSecs > 0 && !runningLatch.await(waitTimeSecs, TimeUnit.SECONDS)) {
+      logger.error("Compactor did not shutdown within " + waitTimeSecs + " seconds");
+    }
+  }
+
+  /**
+   * Compacts the store by copying valid data from the log segments in {@code details} to swap spaces and switching
+   * the compacted segments out for the swap spaces. Returns once the compaction is complete.
+   * @param details the {@link CompactionDetails} for the compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log or if one or more offsets
+   * in the segments to compact are in the journal.
+   * @throws IllegalStateException if the compactor has not been started.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying, commit or cleanup.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void compact(CompactionDetails details) throws InterruptedException, IOException, StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been started");
+    }
+    checkSanity(details);
+    compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, time, details);
+    resumeCompaction();
+  }
+
+  /**
+   * Resumes compaction from where it was left off.
+   * @throws IllegalStateException if the compactor has not been started or if there is no compaction to resume.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying, commit or cleanup.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  void resumeCompaction() throws InterruptedException, IOException, StoreException {
+    if (srcIndex == null) {
+      throw new IllegalStateException("Compactor has not been started");
+    } else if (compactionLog == null) {
+      throw new IllegalStateException("There is no compaction to resume");
+    }
+    runningLatch = new CountDownLatch(1);
+    while (isActive && !compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case PREPARE:
+          compactionLog.markCopyStart();
+          // fall through to COPY
+        case COPY:
+          copy();
+          if (isActive) {
+            compactionLog.markCommitStart();
```

Comment:
When this function is called, compaction cannot be in the COMMIT or CLEANUP phases.